staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go

```diff
@@ -96,8 +96,12 @@ func ResolveCluster(services listersv1.ServiceLister, namespace, id string, port
 	}
 
 	switch {
+	// use service domain for Headless Service
 	case svc.Spec.Type == v1.ServiceTypeClusterIP && svc.Spec.ClusterIP == v1.ClusterIPNone:
-		return nil, fmt.Errorf(`cannot route to service with ClusterIP "None"`)
+		return &url.URL{
+			Scheme: "https",
+			Host:   net.JoinHostPort(fmt.Sprintf("%s.%s.svc", svc.Name, svc.Namespace), fmt.Sprintf("%d", port)),
```

Comment:
Wait... How will coredns resolve this `svc1.ns1.svc`? Our apiserver can run as a container as well.

Reply:
agree in-cluster DNS is not guaranteed to resolve here. that's the whole reason we look up cluster or endpoint IPs in this resolver

Reply:
yes, it's rely on in-cluster DNS service(coredns or kube-dns), and it's not guaranteed also. But using the in-cluster or endpoint IPs is not guaranteed too, because the Kube-Apiserver may run as a container and not a status Pod, as @dixudx said.

Anyway, it's really confused if we set Kube-Apiserver flag `--enable-aggregator-routing=true`, then Kube-Apiserver can proxy using Headless Service endpoints IPs.

Reply:
proxying via headless sounds odd to me. ideally, aggregation should work by redirecting instead of proxying and that's also the whole point we use headless. @lavalamp

Reply:
yes, actually in this case, Kube-Apiserver is more likely doing resolve address and redirect the request to this address. Forgive my poor English ability.. @yue9944882